### PR TITLE
optapy#107: Add ConstraintVerifier API

### LINKED ISSUE #107: Constraint Verifier

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**
Because solution cloning does not occur in the ConstraintVerifier API (because OptaPlanner will not change the given solution/facts), wrapping a fact will cause non-given facts that it reference to have null fields (as wrapping does not go through the entire convert_to_python_like_object process). Thus, after all facts are wrapped, we need to iterate through the values of the reference map (which will contain all facts, including non-given facts) and update their fields from the Python object.

Added test_solve_partial when I was debugging the cause; it already passes with existing solve() code, so the
null field issue only affects the given(facts...) method of ConstraintVerifier. Nonetheless, it is a useful test, so I kept it.

Closes #107

### CODE REVIEW

File: optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java
Comment (triceo) [inline optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java:19]: How public is this API?

I'd prefer it if we didn't use any mutable static structures (shared state) for communication between a user and the program. In this case, I suggest we could put this boolean as a method argument where it is necessary.

File: optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java
Comment (triceo) [inline optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java:59]: My ignorance will show here, bear with me: what is the reference map, where does it come from, how is it used?

File: optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java
Comment (triceo) [inline optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java:75]:
```suggestion
            throw new OptaPyException("A problem occurred when wrapping the python fact (" +
                    PythonWrapperGenerator.getPythonObjectString(fact) + ").\n" 
                    + "Maybe an annotation was passed an incorrect type " +
                    "(for example, @problem_fact_collection_property(str) " +
                    " on a function that returns a list of int).", t);
```

Comment (triceo) [reply optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java:75]: I believe that this is the typical OP way of doing a maybe.

File: optapy-core/src/main/python/test/__init__.py
Comment (triceo) [inline optapy-core/src/main/python/test/__init__.py:12]: What is type checking? Is that some sort of a phase that happens at a different time than at program run time?

File: optapy-core/src/main/python/test/__init__.py
Comment (triceo) [inline optapy-core/src/main/python/test/__init__.py:39]: I'm a bit confused here. I don't see `PythonConstraintVerifier` extend or implement anything, yet there is apparently an overloaded method. What does it overload, from where?

Also, shouldn't `PythonConstraintVerifier` somehow implement OP's `ConstraintVerifier`? How do you ensure that the APIs are in sync?

File: optapy-core/src/main/python/test/__init__.py
Comment (triceo) [inline optapy-core/src/main/python/test/__init__.py:94]: Is it the intention for these classes to be visible to anyone?
What is the policy for this in Python, how does OptaPy specify what is API and what is impl?

File: optapy-core/src/main/python/test/__init__.py
Comment (triceo) [inline optapy-core/src/main/python/test/__init__.py:204]: I understand `...` to mean `pass`. That said, I'm not sure what this means in the context of a method. Is this how Python does abstract methods? Why would it be overloaded then?

File: optapy-core/src/main/python/test/__init__.py
Comment (triceo) [inline optapy-core/src/main/python/test/__init__.py:249]: Why would `times` ever need to be a `str`?

File: optapy-core/src/main/python/test/__init__.py
Comment (triceo) [inline optapy-core/src/main/python/test/__init__.py:485]: I don't understand. Why would we have both `rewards_with` and `rewardsWith`, the latter not being "the Python way"? This makes the API confusing.

File: tests/test_constraint_verifier.py
Comment (triceo) [inline tests/test_constraint_verifier.py:135]: What does this annotation do? What is its equivalent in OptaPlanner?

File: tests/test_constraint_verifier.py
Comment (triceo) [inline tests/test_constraint_verifier.py:171]: Another annotation that I don't think has an equivalent in OptaPlanner.

**Review by triceo — COMMENTED** [review]
Many questions, mostly stemming from my being new to OptaPy and new to Python as well.

Also, is there a follow-up PR planned to introduce tests to OptaPy quickstarts?

File: optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java
Comment (Christopher-Chianelli) [reply optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java:19]: Not public, the user doesn't know it exist. In case you wondering what all the `# noqa` comments are for, it so the IDE typechecker does not raise warnings/errors that it cannot find these packages/classes. That being said, nothing can stop a determined Python programmer from touching internals, but they wouldn't know these internals exist without looking at the source code.

File: optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java
Comment (Christopher-Chianelli) [reply optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java:59]: The reference map is a map from Python ID to Java Object. it is populated during cloning (https://github.com/optapy/optapy/blob/f2d0143fec8052aae4834a9c2d449b2f75ad686c/optapy-core/src/main/python/optaplanner_java_interop.py#L125-L249) and during object translation (https://github.com/optapy/optapy/blob/f2d0143fec8052aae4834a9c2d449b2f75ad686c/jpyinterpreter/src/main/python/python_to_java_bytecode_translator.py#L194-L300). To prevent memory leaks, the map is a "Map with Extras" (that is, a `Map` that delegates to another map EXCEPT for write operations (i.e. put is not delegated)); the base map (which is used only for `get`) is the `jpyinterpreter` reference map (which store things like constants and types, which need to be the same between different functions). The extra map is populated during cloning, which adds the planning solution, planning entities and planning list variables.

File: optapy-core/src/main/python/test/__init__.py
Comment (Christopher-Chianelli) [reply optapy-core/src/main/python/test/__init__.py:12]: `TYPE_CHECKING` is `True` if the code being run by a type checker (an external program used by an IDE), `False` otherwise. It is used here to add type hints for an IDE; These imports will not necessarily be available when the module is imported (because the JVM did not start yet) and thus cab raise an ImportError if we import the package with them. Type stubs are generated for them, so a type checker knows about these classes, even though the implementation is not available.

File: optapy-core/src/main/python/test/__init__.py
Comment (Christopher-Chianelli) [reply optapy-core/src/main/python/test/__init__.py:39]: In Python, `typing.overload` (here imported as `overload`) is used to mark a function definition as a particular function signature for a more broad function. For instance:
```python
@typing.overload
def my_fun(a: int):
    ...

@typing.overload
def my_fun(msg: str):
    ...

@typing.overload
def my_fun(a: int, msg: str):
    ...

def my_fun(a: int | str, msg: str):
    # actual implementation goes here
    ...
```
declares `my_fun` to have three valid signatures: `int`, `str` and `int, str`. Do note Java overloads do not exist in the Python world; a function can only have one signature (which is the most broad one). IDE and type checkers use `typing.overload` to determine return types and parameter types for broad functions.

Comment (Christopher-Chianelli) [reply optapy-core/src/main/python/test/__init__.py:39]: In regards to `PythonConstraintVerifier` somehow reference OP's `ConstraintVerifier`, that can be done with `@JImplements` (which will raise an error if the API's are not in sync). However, there might be an issue if methods are overloaded (for instance, `verifyThat`) (need to verify).

File: optapy-core/src/main/python/test/__init__.py
Comment (Christopher-Chianelli) [reply optapy-core/src/main/python/test/__init__.py:249]: Because `times` might actual be message (i.e. this overload call: https://docs.optaplanner.org/latestFinal/optaplanner-javadoc/org/optaplanner/test/api/score/stream/SingleConstraintAssertion.html#penalizes(java.lang.String) ); this is the delegate for every single overload and thus need to handle each of them

File: optapy-core/src/main/python/test/__init__.py
Comment (Christopher-Chianelli) [reply optapy-core/src/main/python/test/__init__.py:204]: This is how Python declares type signatures for overloaded methods: have a bunch of functions that just do `pass` that declares the signature, followed by the function that implements all of them.

File: optapy-core/src/main/python/test/__init__.py
Comment (Christopher-Chianelli) [reply optapy-core/src/main/python/test/__init__.py:485]: If we want `JImplements` to do API consistency checks, then we need `rewardsWith` (as `OptaPlanner` use `rewardsWith`, so `JImplements` will check for `rewardsWith`). The API being Pythonic (i.e. with `underscores`) was due to comments in this reddit thread (https://www.reddit.com/r/Python/comments/q2euxw/a_new_constraint_solver_in_python_for_the_vehicle/ ; https://github.com/optapy/optapy/issues/22).

File: tests/test_constraint_verifier.py
Comment (Christopher-Chianelli) [reply tests/test_constraint_verifier.py:135]: They are not annotations; they are decorators. Decorators are functions, meaning they can change things and do computation. In this case, it creates a Java class for the Python class, which allows it to be used in Constraint Stream. There no equivalent for OptaPlanner.

File: tests/test_constraint_verifier.py
Comment (Christopher-Chianelli) [reply tests/test_constraint_verifier.py:171]: It does; this is the `ConstraintProvider` interface, which is a functional interface (has only one method, `defineConstraints`).

**Comment by Christopher-Chianelli** [comment]
After this PR Is merged, I will create a PR to the `development` branch of `optapy-quickstarts` (https://github.com/optapy/optapy-quickstarts/tree/development)

File: optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java
Comment (triceo) [reply optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java:19]: Got it. Still, I'd rather we not use shared state. See if you can get rid of it, please.

File: optapy-core/src/main/python/test/__init__.py
Comment (triceo) [reply optapy-core/src/main/python/test/__init__.py:485]: That is unfortunate - we want `rewards_with`, as that is the Python way. At the same time, I don't see why we would have `rewardsWith` as well, that is just confusing. So no `JImplements` for us. :-(

Please remove the `rewardsWith` methods from this API.

File: optapy-core/src/main/python/test/__init__.py
Comment (triceo) [reply optapy-core/src/main/python/test/__init__.py:94]: Ok, no visibility. This Python thing is starting to look like a scripting language.

Anyway... do we need the `Python` prefix in the name of these classes? Seems to me like bloat (obviously it's Python, in Python code). Also, it is camel case, which I understand that the Python people do not like.

**Review by triceo — APPROVED** [review]
LGTM after the remaining comments are resolved.

File: optapy-core/src/main/python/test/__init__.py
Comment (Christopher-Chianelli) [reply optapy-core/src/main/python/test/__init__.py:94]: Classes names are normally PascalCase in Python (i.e. `SomeClass`, not `some_class`). Python Prefix can be removed.

File: optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java
Comment (Christopher-Chianelli) [reply optapy-core/src/main/java/org/optaplanner/optapy/PythonSolver.java:19]: Very hard to remove this shared state, since it is how the `ConstraintProvider` communicates to the `Solver` if it has been successfully FULLY translated to `Java` (and thus can completely avoid going into Python). It is used here: https://github.com/optapy/optapy/blob/db84e20197616175fd606ad371bb63f9caf26565/optapy-core/src/main/python/annotations.py#L664-L678 and here: https://github.com/optapy/optapy/blob/db84e20197616175fd606ad371bb63f9caf26565/optapy-core/src/main/python/optaplanner_api_wrappers.py#L37-L71 ; in short, we cannot know the value until `defineConstraints` is called (since that what compiles all the Python functions), and `definedConstraints` is called during `SolverFactory`/`SolverManager` creation. `PythonSolver` needs a way to read the `ConstraintProvider` instance itself in order to remove that shared state (since `solve` in Python also does not have access to the `ConstraintProvider` that is being used. This field is used to determine what setter to use in ALL domain objects; it will either be a no-op (`(ref, field, value) -> {}`) or it will call CPython (`(ref, field, value) -> CPython.setAttributeOnObject(ref, field, value)`). The objects are wrapped at solve time, and depending on the score function, may need to go to CPython, or can stay in pure java.

**Pull request merged**
